File: frontend/src/lib/api/schemas/code-review.ts
import { z } from "zod/v4"

export const GraphNodeSchema = z.object({
  id: z.string(),
  name: z.string(),
  kind: z.enum(["function", "class", "type", "module", "variable", "export"]),
  filePath: z.string(),
  lineNumber: z.number().optional(),
})

export const GraphEdgeSchema = z.object({
  source: z.string(),
  target: z.string(),
  type: z.enum(["import", "call", "extends", "implements"]),
})

export const GraphDataSchema = z.object({
  nodes: z.array(GraphNodeSchema),
  edges: z.array(GraphEdgeSchema),
  warning: z.string().optional(),
})

export const GraphStatsSchema = z.object({
  built: z.boolean(),
  nodeCount: z.number(),
  edgeCount: z.number(),
  fileCount: z.number(),
  lastBuiltAt: z.string().optional(),
})

export const ImpactResultSchema = z.object({
  changedFiles: z.array(z.string()),
  affectedNodes: z.array(GraphNodeSchema),
  affectedFiles: z.array(z.string()),
  depth: z.number(),
})

export type GraphNode = z.infer<typeof GraphNodeSchema>
export type GraphEdge = z.infer<typeof GraphEdgeSchema>
export type GraphData = z.infer<typeof GraphDataSchema>
export type GraphStats = z.infer<typeof GraphStatsSchema>
export type ImpactResult = z.infer<typeof ImpactResultSchema>

File: backend/schemas/code_review.py
"""Pydantic schemas for the code review dependency graph feature."""

from pydantic import BaseModel


class GraphNode(BaseModel):
    id: str
    name: str
    kind: str  # "function" | "class" | "type" | "module" | "variable" | "export"
    file_path: str
    line_number: int | None = None


class GraphEdge(BaseModel):
    source: str
    target: str
    type: str  # "import" | "call" | "extends" | "implements"


class GraphData(BaseModel):
    nodes: list[GraphNode]
    edges: list[GraphEdge]
    warning: str | None = None


class GraphStats(BaseModel):
    built: bool
    node_count: int
    edge_count: int
    file_count: int
    last_built_at: str | None = None


class ImpactResult(BaseModel):
    changed_files: list[str]
    affected_nodes: list[GraphNode]
    affected_files: list[str]
    depth: int


class BuildRequest(BaseModel):
    full_rebuild: bool = False


class GraphDataRequest(BaseModel):
    mode: str = "full"  # "full" | "focus" | "modified"
    targets: list[str] = []
    depth: int = 2
    kind_filter: list[str] | None = None


class ImpactRequest(BaseModel):
    changed_files: list[str]
    max_depth: int = 3
